File: python/challenges/same_frequency.py
# same_frequency takes two integers and compare the frequency
# should returns false if the frequency is different
#  and returns true when the frequency match
def same_frequency(num1, num2):
    # create list for both numbers
    num1_list = list(str(num1))
    num2_list = list(str(num2))
    # checks if the length
    if len(num1_list) != len(num2_list):
        return False
    # create a dict for storing the first num frequency of digs
    num1_freq = {}

    # loop thru the first num list and add the frequency to the dict
    for dig in num1_list:
        if dig in num1_freq:
            num1_freq[dig] += 1
        else:
            num1_freq[dig] = 1

    # loop thru the second num list
    for dig in num2_list:
        # check if the dig is on the frequency dict and returns false
        if dig not in num1_freq:
            return False
        # else if the frequency value is zero or less returns false
        elif num1_freq[dig] <= 0:
            return False
        # otherwise, decrease the frequency in dict
        else:
            num1_freq[dig] -= 1

    # if all checks pass returns true
    return True

File: javascript/src/challenges/same_frequency.ts
// sameFrequency takes two numbers and comparer
// if the frequency of each digit is same with the other O(n)
function sameFrequency(num1: number, num2: number): boolean {
    // creates variables for storing a array of dig for both numbers
    const num1Dig = String(num1).split('')
    const num2Dig = String(num2).split('')
    // check if the length is different and return false
    if (num1Dig.length !== num2Dig.length) return false

    // creates a object to save the frequency of the first number
    const num1Freq: { [x: string]: number } = {}

    // loop thru the the first number and save the frequency in the object
    for (let i = 0; i < num1Dig.length; i++) {
        const dig = num1Dig[i]
        num1Freq[dig] = num1Freq[dig] + 1 || 1
    }

    // loop thru the the second number
    for (let i = 0; i < num2Dig.length; i++) {
        const dig = num2Dig[i]
        // if the dig is not on frequency object, return false
        if (!num1Freq[dig]) {
            return false
            // otherwise if the frequency is 0 or less return false
        } else if (num1Freq[dig] <= 0) {
            return false
            // otherwise decrease the frequency
        } else {
            num1Freq[dig] -= 1
        }
    }

    // return true if checks pass
    return true
}

export default sameFrequency
